src/backend/resourcemanager/resqueuemanager.c

```diff
@@ -2165,103 +2165,87 @@ void cancelResourceAllocRequest(ConnectionTrack conntrack)
+										 errorbuf);
 }
 
 /* Acquire resource from queue. */
-int acquireResourceFromResQueMgr(ConnectionTrack conntrack)
+int acquireResourceFromResQueMgr(ConnectionTrack  conntrack,
+								 char 			 *errorbuf,
+								 int 			  errorbufsize)
 {
-	int						res			  	= FUNC_RETURN_OK;
-
-	DynResourceQueueTrack   queuetrack	  	= conntrack->QueueTrack;
-
-	if ( queuetrack->ClusterSegNumberMax == 0 )
-	{
-		elog(LOG, "The queue %s has no resource available to run queries.",
-				  queuetrack->QueueInfo->Name);
-		return RESQUEMGR_NO_RESOURCE;
-	}
+	int						res			= FUNC_RETURN_OK;
+	DynResourceQueueTrack	queuetrack	= conntrack->QueueTrack;
 
 	/* Call quota logic to make decision of resource for current query. */
-	res = computeQueryQuota(conntrack);
+	res = computeQueryQuota(conntrack, errorbuf, errorbufsize);
 
 	if ( res == FUNC_RETURN_OK )
 	{
 		if ( conntrack->StatNVSeg == 0 )
 		{
-			int32_t Rmax  = conntrack->SegNum;
-			int32_t RmaxL = conntrack->VSegLimitPerSeg * PRESPOOL->AvailNodeCount;
-			int32_t Rmin  = conntrack->SegNumMin;
-			elog(LOG, "Original quota min seg num:%d, max seg num:%d",
-					  conntrack->SegNumMin,
-					  conntrack->SegNum);
-
-			/* Ensure quota [min,max] is between request [min,max] */
-			int32_t Gmax= conntrack->MaxSegCountFixed;
-			int32_t Gmin= conntrack->MinSegCountFixed;
-
-			if(Gmin==1)
-			{
-				/* case 1 */
-				conntrack->SegNumMin = min(min(Gmax,Rmin),RmaxL);
-				conntrack->SegNum = min(Gmax,RmaxL);
-				if(conntrack->SegNumMin > conntrack->SegNum)
-				{
-					return RESQUEMGR_NO_RESOURCE;
-				}
-			}
-			else if(Gmax == Gmin)
-			{
-				/* case 2 */
-				conntrack->SegNumMin = Gmax;
-				conntrack->SegNum = Gmax;
-				if(Rmax < Gmax)
-				{
-					return RESQUEMGR_NO_RESOURCE;
-				}
-			}
-			else
-			{
-				/* case 3 */
-				conntrack->SegNumMin = min(max(Gmin,Rmin),Gmax);
-				conntrack->SegNum = min(max(min(RmaxL,Gmax),Gmin),Rmax);
-				if(conntrack->SegNumMin > conntrack->SegNum)
-				{
-					return RESQUEMGR_NO_RESOURCE;
-				}
-			}
-
-			elog(LOG, "Query resource expects (%d MB, %lf CORE) x %d ( min %d ) resource.",
-					   conntrack->SegMemoryMB,
-					   conntrack->SegCore,
-					   conntrack->SegNum,
-					   conntrack->SegNumMin);
-
+			/*------------------------------------------------------------------
+			 * The following logic consider the actual resource requirement from
+			 * dispatcher based on table size, workload, etc. The requirement is
+			 * described by (MinSegCountFixed, MaxSegCountFixed). The requirement
+			 * can be satisfied only when there is a non-empty intersect between
+			 * (MinSegCountFixed, MaxSegCountFixed) and (SegNumMin, SegNum).
+			 *------------------------------------------------------------------
+			 */
+			conntrack->SegNumMin =
+				conntrack->MaxSegCountFixed < conntrack->SegNumMin ?
+				conntrack->MinSegCountFixed :
+				max(conntrack->SegNumMin, conntrack->MinSegCountFixed);
+
+			conntrack->SegNum = min(conntrack->SegNum,
```

Comment:
can you explain the meaning of MinSegCountFixed, MaxSegCountFixed, MinSegCountFixed and MaxSegCountFixed?